docs/AQL/grammar/AqlLexer.g4

```diff
@@ -0,0 +1,299 @@
+//
+//  description:  ANTLR4 lexer grammar for Archetype Query Language (AQL)
+//  authors:      Sebastian Iancu, Code24, Netherlands
+//                Teun van Hemert, Nedap, Netherlands
+//                Thomas Beale, Ars Semantica UK, openEHR Foundation Management Board
+//  contributors: This version of the grammar is a complete rewrite of previously published antlr3 grammar,
+//                based on current AQL specifications in combination with other grammars from several AQL implementations.
+//                The openEHR Foundation would like to recognise the following people for their contributions:
+//                  - Chunlan Ma & Heath Frankel, Ocen Health Systems, Australia
+//                  - Bostjan Lah, Better, Slovenia
+//                  - Christian Chevalley, EHRBase, Germany
+//                  - Michael Böckers, Nedap, Netherlands
+//  support:      openEHR Specifications PR tracker <https://specifications.openehr.org/releases/QUERY/open_issues>
+//  copyright:    Copyright (c) 2020 openEHR Foundation
+//  license:      Apache 2.0 License <http://www.apache.org/licenses/LICENSE-2.0.html>
+//
+
+lexer grammar AqlLexer;
+
+channels {
+    COMMENT_CHANNEL
+}
+
+// SKIP
+WS: [ \t\r\n]+ -> skip;
+UNICODE_BOM: (
+    '\uEFBBBF' // UTF-8 BOM
+    | '\uFEFF' // UTF16_BOM
+    | '\u0000FEFF' // UTF32_BOM
+    ) -> skip;
+COMMENT: (
+    SYM_DOUBLE_DASH ' ' ~[\r\n]* ('\r'? '\n' | EOF)
+    | SYM_DOUBLE_DASH ('\r'? '\n' | EOF)
+    ) -> channel(COMMENT_CHANNEL);
+
+// Keywords
+// Common Keywords
+SELECT: S E L E C T ;
+AS: A S ;
+FROM: F R O M ;
+WHERE: W H E R E ;
+ORDER: O R D E R ;
+BY: B Y ;
+DESC: D E S C ;
+DESCENDING: D E S C E N D I N G ;
+ASC: A S C ;
+ASCENDING: A S C E N D I N G ;
+LIMIT: L I M I T ;
+OFFSET: O F F S E T ;
+// other keywords
+DISTINCT: D I S T I N C T ;
+EHR : E H R ;
```

Comment:
I don't think this keyword is needed - and it should not normally be a keyword. I can't see it used anywhere else anyway.

Reply:
is is used by classExprOperand parser rule
```
classExprOperand
    : TYPE_ID variable=VARIABLE_ID? (SYM_LEFT_BRACKET archetypePredicate SYM_RIGHT_BRACKET)?                    #classExpression // RM_TYPE_NAME variable [archetype_id]
    | VERSIONED_OBJECT variable=VARIABLE_ID? (SYM_LEFT_BRACKET standardPredicate SYM_RIGHT_BRACKET)?            #versionedClassExpr
    | EHR variable=VARIABLE_ID? (SYM_LEFT_BRACKET standardPredicate SYM_RIGHT_BRACKET)?                         #ehrClassExpr
    | VERSION variable=VARIABLE_ID? (SYM_LEFT_BRACKET (standardPredicate|versionPredicate) SYM_RIGHT_BRACKET)?  #versionClassExpr
    ;
```

EHR class there can only be associated with a standard predicate. I thought we might one to keep this generic, not mentioning any RM class name, but I have trouble finding the right definition for associated predicate.

Perhaps we could optimize this (make it more compact, less alternatives), but I was not sure if it is a good idea.

Reply:
Ah I see. What I would usually do here is to keep the grammar generic, and use a validation phase to check for validity relating to specific types. I'm not sure why we need VERSIONED_OBJECT - is there a requirement for that? So you could do this.

```
versionPredicate:
      LATEST_VERSION
    | ALL_VERSIONS 
    | standardPredicate
    ;

classExprOperand
    : TYPE_ID variable=VARIABLE_ID? (SYM_LEFT_BRACKET pathPredicate SYM_RIGHT_BRACKET)?
    | VERSION variable=VARIABLE_ID? (SYM_LEFT_BRACKET versionPredicate SYM_RIGHT_BRACKET)?  #versionClassExpr
    ;
```

The validator would read some rules (which could even be hardwired in a rules class, in a simple query evaluator), that allow only certain predicates with the root object type ('EHR' in the RM, something else in another model). I had a look at the predicate rules - I don't understand all of them yet, but will review properly in another pass.

Reply:
done in https://github.com/openEHR/specifications-QUERY/pull/15/commits/ae50c5194bf81295e044ae011d60688d9c1a260a